eudemo/eudemo/comp_managers.py

```diff
@@ -52,12 +61,115 @@ def xz_boundary(self) -> BluemiraWire:
+        vvts_void_name = f"{VVTSBuilder.VOID} 1"
+        vvts_target_shape = vv_xyz.get_component(vvts_target_name).shape
+        vvts_target_void = vv_xyz.get_component(vvts_void_name).shape
+
+        cts_xyz = cts.get_component("xyz")
+        cr_xyz = cts_xyz.get_component("Sector 1")
+        cts_target_name = f"{CryostatTSBuilder.CRYO_TS} 1"
+        cts_void_name = f"{CryostatTSBuilder.VOID} 1"
+        cts_target_shape = cr_xyz.get_component(cts_target_name).shape
+        cts_target_void = cr_xyz.get_component(cts_void_name).shape
+
+        if isinstance(ports, Component):
+            ports = [ports]
+
+        tool_voids = []
+        new_shape_pieces = []
+        for port in ports:
+            port_xyz = port.get_component("xyz")
+            tool_shape = port_xyz.get_component(port.name).shape
+            tool_void = port_xyz.get_component(port.name + " voidspace").shape
+            tool_voids.append(tool_void)
+            result_pieces = pipe_pipe_join(
+                vvts_target_shape, vvts_target_void, tool_shape, tool_void
+            )
+            # Assume the body is the biggest piece
+            result_pieces.sort(key=lambda solid: -solid.volume)
+            vvts_target_shape = result_pieces[0]
+            new_shape_pieces.extend(result_pieces[1:])
+
+        final_shape = boolean_fuse([vvts_target_shape] + new_shape_pieces)
+        final_void = boolean_fuse([vvts_target_void] + tool_voids)
+
+        temp = boolean_cut(final_shape, cts_target_shape)
+        temp.sort(key=lambda solid: -solid.volume)
+        final_shape = temp[0]
+
+        cts_target_shape = boolean_cut(cts_target_shape, tool_voids)[0]
+
+        vvts_sector_body = PhysicalComponent(vvts_target_name, final_shape)
+        vvts_sector_void = PhysicalComponent(
+            vvts_void_name, final_void, material=Void("vacuum")
+        )
+        apply_component_display_options(vvts_sector_body, color=BLUE_PALETTE["TS"][0])
+        apply_component_display_options(vvts_sector_void, color=(0, 0, 0))
+
+        cts_sector_body = PhysicalComponent(cts_target_name, cts_target_shape)
+        cts_sector_void = PhysicalComponent(
+            cts_void_name, cts_target_void, material=Void("vacuum")
+        )
+        apply_component_display_options(cts_sector_body, color=BLUE_PALETTE["TS"][0])
+        apply_component_display_options(cts_sector_void, color=(0, 0, 0))
+
+        # Orphan and kill old shapes
+        vvts_xyz.parent = None
+        del vvts_xyz
+        cts_xyz.parent = None
+        del cts_xyz
+        # Replace xyz components
+        Component(
+            "xyz",
+            children=[
+                Component("Sector 1", children=[vvts_sector_body, vvts_sector_void])
+            ],
+            parent=vvts,
+        )
+        Component(
+            "xyz",
+            children=[
+                Component("Sector 1", children=[cts_sector_body, cts_sector_void])
+            ],
+            parent=cts,
+        )
```

Comment:
I've had a bit of a go a lying this out slightly differently, let me know what you think

```suggestion
    @staticmethod
    def _join_ports_to_vvts(
        ports: Union[Component, List[Component]],
        vvts_target_shape: BluemiraSolid,
        vvts_target_void: BluemiraSolid,
        cts_target_shape: BluemiraSolid,
    ) -> Tuple[BluemiraSolid, List[BluemiraSolid]]:
        """Joins the ports to the vvts and cuts at the cryo_ts"""
        if isinstance(ports, Component):
            ports = [ports]

        tool_voids = []
        new_shape_pieces = []

        for port in ports:
            port_xyz = port.get_component("xyz")
            tool_shape = port_xyz.get_component(port.name).shape
            tool_voids.append(port_xyz.get_component(f"{port.name} voidspace").shape)
            result_pieces = pipe_pipe_join(
                vvts_target_shape, vvts_target_void, tool_shape, tool_voids[-1]
            )
            # Assume the body is the biggest piece
            result_pieces.sort(key=lambda solid: -solid.volume)
            vvts_target_shape = result_pieces[0]
            new_shape_pieces.extend(result_pieces[1:])

        final_shapes = boolean_cut(
            boolean_fuse([vvts_target_shape] + new_shape_pieces), cts_target_shape
        )
        final_shapes.sort(key=lambda solid: -solid.volume)
        return final_shapes[0], tool_voids

    def add_ports(self, ports: List[Component], n_TF: int):
        """
        Add ports to the thermal shield
        """

        cts = self.cryostat_thermal_shield()
        cts_target_name = f"{CryostatTSBuilder.CRYO_TS} 1"
        cts_void_name = f"{CryostatTSBuilder.VOID} 1"
        cts_xyz = cts.get_component("xyz").get_component("Sector 1")
        cts_target_shape = cts_xyz.get_component(cts_target_name).shape

        vvts = self.vacuum_vessel_thermal_shield()
        vvts_target_name = f"{VVTSBuilder.VVTS} 1"
        vvts_void_name = f"{VVTSBuilder.VOID} 1"
        vvts_xyz = vvts.get_component("xyz").get_component("Sector 1")
        vvts_target_void = vvts_xyz.get_component(vvts_void_name).shape
        vvts_target_shape, tool_voids = self._join_ports_to_vvts(
            ports,
            vvts_xyz.get_component(vvts_target_name).shape,
            vvts_target_void,
            cts_target_shape,
        )

        # Replace xyz components
        for comp, void_name, void, body_name, body in (
            (
                vvts,
                vvts_void_name,
                boolean_fuse([vvts_target_void] + tool_voids),
                vvts_sector_void,
                vvts_target_name,
                vvts_target_shape,
            ),
            (
                cts,
                cts_void_name,
                cts_xyz.get_component(cts_void_name).shape,
                cts_target_name,
                boolean_cut(cts_target_shape, tool_voids)[0],
            ),
        ):
            sector_body = PhysicalComponent(body_name, body)
            sector_void = PhysicalComponent(void_name, void, material=Void("vacuum"))

            apply_component_display_options(sector_body, color=BLUE_PALETTE["TS"][0])
            apply_component_display_options(sector_void, color=(0, 0, 0))

            # Orphan and kill old shapes
            comp.get_component("xyz").parent = None

            Component(
                "xyz",
                children=[Component("Sector 1", children=[sector_body, sector_void])],
                parent=comp,
            )
```

Reply:
with a few variable name tweaks it runs for me, but breaks. I will have a go at cleaning it up